lib/linguist/heuristics.yml

```diff
@@ -156,9 +156,9 @@ disambiguations:
 - extensions: ['.ice']
   rules:
   - language: Slice
-    pattern: '^\s*(#\s*(include|if[n]def|pragma)|module\s+[A-Za-z][_A-Za-z0-9]*)'
+    pattern: '\A(?!\s*[{\[])'
```

Comment:
Isn't this the exact opposite of the JSON regex? Do we need both?

Reply:
We *should* be able to get by with only `'\A(?!\s*[{\[])'` for Slice. That, or remove the pattern altogether and use an open-ended fallback to Slice with JSON checked first. Which is probably clearer in the long-run. I'm still puzzled as to why the tests are failing. 😕

Also, I can't even see the diff I'm reviewing. 😀 Scrollbar isn't going away:
![a](https://user-images.githubusercontent.com/2346707/51082811-a856b780-1762-11e9-8192-3dd8efeb28fa.png)

Reply:
> remove the pattern altogether and use an open-ended fallback to Slice with JSON checked first

addressed in 9c2302a

> I'm still puzzled as to why the tests are failing. 

I've put back smallest JSON sample for .ice in bd3b268 and one can see that it's not heuristics but only a Classifier test that is failing `TestClassifier#test_classify_ambiguous_languages`.

This [Classifier test](https://github.com/github/linguist/blob/58193319b3a1daa8fae0fec897029137a958568e/test/test_classifier.rb#L46) seems to be designed to look for all the languages that are ambiguous by extension (JSON and Slice in `.ice` case) and then assume `Classifier.classify` would be able to distinguish those without using heuristics strategy that we are adjusting in here.

This seems to be not possible, only based on samples that we have for this pair of languages 

8kb for Slice (that also uses `{} and []`) VS 216kb total for JSON:
```
      JSON =   -600.095 +  -4.551 =   -604.646
     Slice =   -565.507 +  -6.949 =   -572.456
```

instead of adding bigger JSON sample, I tried adding +16kb of Slice one and it seems to _almost_ work

24kb Slice VS  216kb for JSON:
```
      JSON =   -600.095 +  -4.552 =   -604.647
     Slice =   -596.006 +  -6.544 =   -602.551
```

Reply:
+12kb of Slice samples (total +32kb VS +96kb in previous approach) seems to do the trick

```
      JSON =   -600.095 +  -4.552 =   -604.647
     Slice =   -624.850 +  -6.257 =   -631.108
```

🎉

Reply:
Good work! Wait, so, the heuristics strategy *doesn't* always run??

Reply:
It does not in `TestClassifier`, as this test directly uses only a low-level Baesian classifier API `Classifier.classify()`

I belive it may be the assumptions of this test that is too strong - for some reason it _assumes_ that all colliding extensions can be disambiguated by Baesian classifier directly. 
When in Linguist \w multiple strategies this classifier should be able to disambiguate only the cases that are _not_ covered by heuristics already.